metric-forecasting/metric_streaming.py

```diff
@@ -61,6 +66,62 @@
     "disk_usage",
 ]  ## TODO: default metrics and their queries should be configured in a file.
 
+GAUGE_DICT = dict()
+MAD_DICT = dict()
+PROMETHEUS_CUSTOM_QUERIES = {
+    'cpu_usage': '100 * (1- (avg(irate(node_cpu_seconds_total{mode="idle"}[5m]))))',
+    'memory_usage': '100 * (1 - sum(node_memory_MemAvailable_bytes) / sum(node_memory_MemTotal_bytes))',
+    'disk_usage': '(sum(node_filesystem_size_bytes{device!~"rootfs|HarddiskVolume.+"})- sum(node_filesystem_free_bytes{device!~"rootfs|HarddiskVolume.+"})) / sum(node_filesystem_size_bytes{device!~"rootfs|HarddiskVolume.+"}) * 100 ',
+}
+COLUMNS_LIST = [
+    'is_anomaly',
+    'alert_score',
+    'alert_len',
+    'alert_id'
+    'is_alert',
+    'y',
+    'yhat',
+    'yhat_lower',
+    'yhat_upper',
+]
+
+@app.on_event("startup")
+async def startup_event():
+    logger.info("loading history on webserver startup")
+    for metric_name in PROMETHEUS_CUSTOM_QUERIES.keys():
+        MAD_DICT[metric_name] = MetricAnomalyDetector(metric_name)
+        GAUGE_DICT[metric_name] = Gauge(metric_name, metric_name, ['value_type'])
+        await load_history_data(metric_name, MAD_DICT[metric_name])
+
+@app.get("/")
+@app.get("/metrics")
+async def get_metrics():
+    metrics_payloads = []
+    for metric_name, query in PROMETHEUS_CUSTOM_QUERIES.items():
+        current_data = prom.custom_query(query=query)
+        prediction = MAD_DICT[metric_name].run(current_data)
+        if prediction:
+            metrics_payloads.append(prediction)
+            for column in COLUMNS_LIST:
+                if prediction[column] is not None:
+                    GAUGE_DICT[metric_name].labels(value_type=column).set(prediction[column])
```

Comment:
do we need the GAUGE_DICT ? I don't find it being  used anywhere else. is it reserved for Thanos ?

Reply:
Yeah, the Gauge objects in the dictionary are instantiated in the startup
```
async def startup_event():
    logger.info("loading history on webserver startup")
    for metric_name in PROMETHEUS_CUSTOM_QUERIES.keys():
        MAD_DICT[metric_name] = MetricAnomalyDetector(metric_name)
        GAUGE_DICT[metric_name] = Gauge(metric_name, metric_name, ['value_type'])
        await load_history_data(metric_name, MAD_DICT[metric_name])
```
Then the method that actually writes out the prometheus metrics makes use of them
```
generate_latest(REGISTRY).decode('utf-8')
```

It's all part of the prometheus-client library